crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,355 @@
+//!       do nothing.
+//!     - Otherwise proceed.
+//! - Check if agent is on a cool-down list of unresponsive agents.
+//! - Dispatch request for op id from agent to transport module.
+//! - If agent is unresponsive, put them on cool-down list.
+//! - Re-send requested ([OpId], [AgentId]) to the internal channel again. It will be removed
+//!   from the list of ops to fetch if it is received in the meantime, and thus prevent a redundant
+//!   fetch request.
+//!
+//! ### Incoming op task
+//!
+//! - Incoming op is written to the data store.
+//! - Once persisted successfully, op is removed from the data object.
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use bytes::BufMut;
+use kitsune2_api::{
+    builder,
+    config::ModConfig,
+    fetch::{DynFetch, DynFetchFactory, Fetch, FetchFactory},
+    peer_store,
+    transport::Transport,
+    AgentId, BoxFut, K2Result, OpId, SpaceId,
+};
+use tokio::{
+    sync::mpsc::{channel, Receiver, Sender},
+    task::JoinHandle,
+};
+
+const MOD_NAME: &str = "Fetch";
+
+/// A production-ready fetch module.
+#[derive(Debug)]
+pub struct CoreFetchFactory {}
+
+impl CoreFetchFactory {
+    /// Construct a new CoreFetchFactory.
+    pub fn create() -> DynFetchFactory {
+        Arc::new(Self {})
+    }
+}
+
+impl FetchFactory for CoreFetchFactory {
+    fn default_config(
+        &self,
+        config: &mut kitsune2_api::config::Config,
+    ) -> K2Result<()> {
+        config.add_default_module_config::<CoreFetchConfig>(
+            MOD_NAME.to_string(),
+        )?;
+        Ok(())
+    }
+
+    fn create(
+        &self,
+        builder: Arc<builder::Builder>,
+        space_id: SpaceId,
+        peer_store: peer_store::DynPeerStore,
+        transport: DynTransport,
+    ) -> BoxFut<'static, K2Result<DynFetch>> {
+        Box::pin(async move {
+            let config = builder.config.get_module_config(MOD_NAME)?;
+            let out: DynFetch = Arc::new(CoreFetch::new(
+                config, space_id, peer_store, transport,
+            ));
+            Ok(out)
+        })
+    }
+}
+
+/// Configuration parameters for [CoreFetchFactory].
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoreFetchConfig {
+    /// How many parallel op fetch requests can be made at once. Default: 2.  
+    pub parallel_request_count: u8,
+    /// Duration in ms to keep an unresponsive agent on the cool-down list. Default: 120_000.
+    pub cool_down_interval_ms: u64,
+}
+
+impl Default for CoreFetchConfig {
+    fn default() -> Self {
+        Self {
+            parallel_request_count: 2,
+            cool_down_interval_ms: 120_000,
+        }
+    }
+}
+
+impl ModConfig for CoreFetchConfig {}
+
+// TODO: Temporary trait object of a transport module to facilitate unit tests.
+type DynTransport = Arc<dyn Transport>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
```

Comment:
If the mutex isn't going to be at this level, there's no need for Inner. You could implement Fetch on Inner. Then rename Inner to CoreFetch.

Reply:
Yep, simpler.

Reply:
https://github.com/holochain/kitsune2/pull/35/commits/e0c11665a8f6cb3518bf103c72f61fadaaffb418